--- tmp/start.ini	2010-09-22 16:48:02.570074526 +1000
+++ start.ini	2010-09-22 16:50:13.934074094 +1000
@@ -42,7 +42,8 @@
 # for a full listing do
 #   java -jar start.jar --list-options
 #-----------------------------------------------------------
-OPTIONS=Server,jsp,jmx,resources,websocket,ext
+OPTIONS=Server,jsp,jmx,resources,websocket,ext,jta,plus,jdbc,annotations
+
 #-----------------------------------------------------------
 
 
@@ -53,10 +54,13 @@
 #-----------------------------------------------------------
 #etc/jetty-jmx.xml
 etc/jetty.xml
+etc/jetty-hightide.xml
 # etc/jetty-ssl.xml
 # etc/jetty-requestlog.xml
 etc/jetty-deploy.xml
+etc/jetty-plus.xml
 etc/jetty-webapps.xml
 etc/jetty-contexts.xml
 etc/jetty-testrealm.xml
+etc/jetty-jaas.xml
 #===========================================================
