Index: org/apache/taglibs/standard/tag/common/sql/DataSourceWrapper.java
===================================================================
--- org/apache/taglibs/standard/tag/common/sql/DataSourceWrapper.java	(revision 5)
+++ org/apache/taglibs/standard/tag/common/sql/DataSourceWrapper.java	(working copy)
@@ -132,6 +132,19 @@
     public synchronized void setLogWriter(PrintWriter out) throws SQLException {
         throw new SQLException(Resources.getMessage("NOT_SUPPORTED"));
     }
+    
+    /**
+    * Always return false.
+    */
+    public boolean isWrapperFor(Class<?> iface) throws SQLException {
+        return false;
+    }
+    /**
+    * Always throws a SQLException. Not supported.
+    */
+    public <T> T unwrap(Class<T> iface) throws SQLException {
+        throw new SQLException(Resources.getMessage("NOT_SUPPORTED"));
+    }    
 
 
 }
