Index: EmbeddedServletOptions.java
===================================================================
--- EmbeddedServletOptions.java	(revision 4)
+++ EmbeddedServletOptions.java	(working copy)
@@ -533,8 +533,8 @@
         // START PWC 1.2 6311155
         String sysClassPath = config.getInitParameter(
                                     "com.sun.appserv.jsp.classpath");
-        if (sysClassPath != null)
-            this.sysClassPath = sysClassPath;
+        this.sysClassPath = sysClassPath!=null ? sysClassPath : System.getProperty("java.class.path");
+            
         // END PWC 1.2 6311155
 
 	/*
